Implement new COM interop API for RCW/CCW creation/management

AaronRobinsonMSFT opened the pull request:
Implementation of the proposed API found at https://github.com/dotnet/runtime/issues/1845.

All new tests run and pass on Windows running under `COMPlus_GCStress=0xC`.

In order to allow code generators to make performant RCW/CCW implementations, the [`NativeCallableAttribute` should be made public](https://github.com/dotnet/runtime/issues/32462) and the [C# function pointers proposal](https://github.com/dotnet/csharplang/blob/master/proposals/function-pointers.md) should be implemented.

**TODO:**
- [x] Complete `Peg`/`Unpeg` API implementation.

- [x] Remove all new `TODO`s.

- [x] Finish E2E testing provided by the Mock Tracker Runtime project.

- [x] Add testing for skip cache scenarios when creating managed wrappers.

- [x] Handle build issues on non-Windows platforms.

- [x] Remove shutdown logic once https://github.com/dotnet/runtime/pull/32381 is merged in.

- [x] Add logging to the runtime side of the interoplib library.

- [x] Investigate how the following COM APIs work with this API.
  - ~~[`Marshal.ReleaseComObject()`](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.marshal.releasecomobject)~~ - Properly handled.
  - ~~[`Marshal.FinalReleaseComObject()`](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.marshal.finalreleasecomobject)~~ - Properly handled.
  - ~~[`Marshal.GetIUnknownForObject()`](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.marshal.getiunknownforobject)~~ - Calling this on a wrapper object created by this new API will passively create an old style CCW. Deferring any changes until customer reported.

**Future issues to file:**

* ~~Share additional code between existing RCW/CCW support and new API.~~

* Update `Marshal.GetIUnknownForObject()` and `Marshal.GetObjectForIUnknown()` to check and possible use the globally registered `ComWrappers` instance - see https://github.com/dotnet/runtime/pull/32091#issuecomment-592295188.

  * See https://github.com/dotnet/runtime/issues/33329

* Consider removing the existing built-in WinRT support - see https://github.com/dotnet/runtime/pull/32091#discussion_r377433619. 

  * See https://github.com/dotnet/runtime/issues/33330.

/cc @jkotas @jkoritzinsky @scottj1s @davidwrighton

AaronRobinsonMSFT committed:
Add API to SPCL.

AaronRobinsonMSFT committed:
Revert previous work.

AaronRobinsonMSFT committed:
Add API in stub form.

AaronRobinsonMSFT committed:
Implement RuntimeHelpers.AllocateTypeAssociatedMemory()

AaronRobinsonMSFT committed:
Add tests for RuntimeHelpers.AllocateTypeAssociatedMemory

AaronRobinsonMSFT committed:
Update interop lib import/export API headers

AaronRobinsonMSFT committed:
Implement the ComWrappers.GetIUnknownImpl() API

AaronRobinsonMSFT committed:
Add project for testing ComWrappers API.
Add test for the ComWrappers.GetIUnknownImpl() API.

AaronRobinsonMSFT committed:
Stub out remaining ComWrapper QCall APIs

AaronRobinsonMSFT committed:
Add pseudo implementation and test for ComWrappers.RegisterForReferenceTrackerHost().

AaronRobinsonMSFT committed:
Add and validate metasigs for invoking static methods in ComWrappers from within the runtime.

AaronRobinsonMSFT committed:
Implement the runtime interation for managed object wrapper creation.
Update InteropSyncBlockInfo to hold onto wrapper instance.

AaronRobinsonMSFT committed:
Write test for creation of managed object wrapper with tracker support.

AaronRobinsonMSFT committed:
Make managed object wrappers persistent until the associated managed object is finalized.
Remove the CreateComInterfaceFlags.IgnoreCache

AaronRobinsonMSFT committed:
Update export namespaces in InteropLib

AaronRobinsonMSFT committed:
Surround UNREACHABLE() with braces because of how it is defined.

AaronRobinsonMSFT committed:
Remove unnecessary GCProtect sections.
Allocate External Object Context cache.

AaronRobinsonMSFT committed:
Retain the External Object Context in the SyncBlock and handle cleanup.

AaronRobinsonMSFT committed:
Rename ExtObjCxtIterator to RuntimeCallContext.
Add API for creating a Reference between an External Object and a MOW.
Update the existing RCWRefCache to work with the new InteropLib API.

AaronRobinsonMSFT committed:
Change the CreateReference() API to an informing style API (i.e. FoundReferencePath).
  - This permits the runtime to make the decision on if the connection is meaningful.

AaronRobinsonMSFT committed:
Make global pegging field on RCWWalker public.

AaronRobinsonMSFT committed:
Place all exported functions into a single compilation unit for InteropLib.
Add Begin/End reference tracking APIs.
Expose Global Pegging value to InteropLib.

AaronRobinsonMSFT committed:
Add new API to make object release requests.

AaronRobinsonMSFT committed:
Address object release issues.

AaronRobinsonMSFT committed:
Update public API

AaronRobinsonMSFT committed:
Add support for checking if an external object implements the
  IReferenceTracker interface from the runtime side.

AaronRobinsonMSFT committed:
Add comment about IEnumerable optimization.

AaronRobinsonMSFT committed:
Delay load ole32.dll to get RoGetAgileReference using GetProcAddress.

AaronRobinsonMSFT committed:
TEST - WIP

AaronRobinsonMSFT committed:
Reduce uses of GetReferenceTracker() - unsafe function.
Update CreateWrapperForExternal() to return a result struct with
  extra details.

AaronRobinsonMSFT committed:
Validation bug for converting to MOW.

AaronRobinsonMSFT committed:
Move the ComInterfaceEntry to the ABI namespace.

AaronRobinsonMSFT committed:
Mock Tracker Runtime impl

jkotas commented on src/coreclr/src/vm/ceemain.cpp:
```diff
@@ -188,7 +188,7 @@
 #include "runtimecallablewrapper.h"
 #include "notifyexternals.h"
 #include "mngstdinterfaces.h"
-#include "rcwwalker.h"
```
Is the RCW walker going to be deleted as part of this?

AaronRobinsonMSFT replied:
That is still TBD. At present the answer is no. The current plan would be to do the removal for .NET 6 or .NET 5.1 instead so we could potentially start removal as soon as the snap from .NET 5 occurs.

jkotas commented on src/coreclr/src/interop/inc/interoplib.h:
```diff
@@ -0,0 +1,96 @@
+
+        // Register a runtime implementation callback in the Reference Tracker Host scenario.
+        // Returns true if registration succeeded, otherwise false.
+        bool RegisterReferenceTrackerHostRuntimeImpl(_In_ OBJECTHANDLE objectHandle) noexcept;
```
This is a singleton anyway. The managed reference can be just a field in C#. I do not think it needs to be wired all the way into the interop lib just so that it can be passed all the way back.

AaronRobinsonMSFT replied:
I typically try to avoid using `NULL` as anything other than a predicate invariant which is why I passed the object around. However, this seems like a reasonable idea to avoid creating yet _another_ object handle.

AaronRobinsonMSFT replied:
Continuing the thought process here. Letting `NULL` through does makes some of these code paths less rigorous since now `NULL` is permitted and the state is more magical. Is there a strong desire to remove this registration? I am not against it but it does loosen so many of the contracts and increases the mental/testing complexity of understanding the system.

jkotas replied:
I meant to delete the OBJECTHANDLE argument. I did not meant to pass NULL.

AaronRobinsonMSFT replied:
Agreed. I can remove the registration, but the usage of this value originates from within the InteropLib via a call from Jupiter. Which means I would need to pass null from there all the way back up to indicate use the static field on the `ComWrappers` rather than being provided one.

AaronRobinsonMSFT replied:
Hmmmm perhaps there is another interpretation for your statement. Is there an invalid `OBJECTREF` value that I can use as an invariant to use?

AaronRobinsonMSFT replied:
Done.

jkotas commented on src/coreclr/src/vm/ceemain.cpp:
```diff
@@ -1188,8 +1188,7 @@ void InnerCoEEShutDownCOM()
 
-    // Release IJupiterGCMgr *
-    RCWWalker::OnEEShutdown();
+    Interop::OnRuntimeShutdown();
```
Is the shutdown doing anything important - what is not going to now work if we delete this?

We strongly prefer to avoid doing work during shutdown.

AaronRobinsonMSFT replied:
Minor cleanup. I can definitely avoid it without much issue. The main thing that is occurring is the release of the tracker runtime manager (i.e. `IReferenceTrackerManager`). This is the contract that the Jupiter runtime would provide us during our initial handshake.

jkotas replied:
We do not suspect other threads during shutdown. There is a chance that other thread may be using this is running, unless there are locks etc. to prevent it.

AaronRobinsonMSFT replied:
>There is a chance that other thread may be using this is running, unless there are locks etc. to prevent it.

Only managed threads and the GC thread use the set value. Can either of those types of threads be running when this function is called?

AaronRobinsonMSFT replied:
What a disaster. The `InnerCoEEShutDownCOM()` is no longer called. @jkotas thoughts on removing the entire thing in this PR or in another?

AaronRobinsonMSFT replied:
I have opted to remove it in a new PR. I will update this one as soon as the removal is merged in - see https://github.com/dotnet/runtime/pull/32381.

jkotas commented on src/coreclr/src/interop/agilereferences.cpp:
```diff
@@ -0,0 +1,56 @@
+
+#include "comwrappers.h"
+
+#if (NTDDI_VERSION >= NTDDI_WINBLUE)
```
I would just bump WINVER as high as necessary so that we do not need this.

AaronRobinsonMSFT replied:
I would prefer to avoid the version bump. I will simple forward declare it in all circumstances.

MichalStrehovsky commented on src/coreclr/src/interop/comwrappers.cpp:
```diff
@@ -0,0 +1,662 @@
+    DisconnectTracker();
+
+#ifdef _DEBUG
+    _sentinal = DeadContextSentinal;
```
Nit: Should all of the `sentin*a*l`  be spelled `sentin*e*l` in this file?

AaronRobinsonMSFT replied:
Sigh... Thanks @MichalStrehovsky!

lpereira commented on src/coreclr/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs:
```diff
@@ -264,6 +264,28 @@ internal static unsafe bool ObjectHasComponentSize(object obj)
+            if (size < 0)
+                throw new ArgumentOutOfRangeException(nameof(size));
+
+            return AllocateTypeAssociatedMemoryInternal(new QCallTypeHandle(ref rt), (uint)size);
```
If `AllocateTypeAssociatedMemoryInternal()` takes a `uint size`, can this also take an unsigned integer and get rid of the if/throw above?

AaronRobinsonMSFT replied:
Public APIs should typically be [`CLSCompliant`](https://docs.microsoft.com/dotnet/api/system.clscompliantattribute) and setting this as an unsigned int would break that.

AaronRobinsonMSFT committed:
Forward declare Win32 API.

AaronRobinsonMSFT committed:
Spelling mistake.

AaronRobinsonMSFT committed:
Move registration of global ComWrappers to be entirely in managed code.

jkotas commented on src/coreclr/src/dlls/mscorrc/mscorrc.rc:
```diff
@@ -919,6 +919,7 @@ BEGIN
     IDS_EE_ATTEMPT_TO_CREATE_GENERIC_CCW    "Generic types cannot be marshaled to COM interface pointers."
     IDS_EE_ATTEMPT_TO_CREATE_NON_ABSTRACT_CCW "Types with non-abstract methods cannot be marshaled to COM interface pointers."
     IDS_EE_COMIMPORT_METHOD_NO_INTERFACE    "Method '%1' in ComImport class '%2' must implement an interface method."
+    IDS_EE_RESET_REFERENCETRACKERHOST_CALLBACKS "Attempt to update previously set default ReferenceTrackerHost callbacks."
```
This error message is not needed on the unmanaged size anymore

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.cs:
```diff
@@ -0,0 +1,224 @@
+        /// <summary>
+        /// Globally registered instance of the ComWrappers class.
+        /// </summary>
+        private static ComWrappers? s_GlobalInstance;
```
```suggestion
        private static ComWrappers? s_globalInstance;
```

AaronRobinsonMSFT committed:
Remove unused message.

AaronRobinsonMSFT committed:
Add error message to managed resources.
Variable name feedback.

AaronRobinsonMSFT committed:
Implement Peg/Unpeg

AaronRobinsonMSFT committed:
Add additional testing.

AaronRobinsonMSFT committed:
Test ignore cache logic.

AaronRobinsonMSFT committed:
Change EnsureActiveWrapperAndAddRef() API to take an indirect OBJECTREF
  and only create an OBJECTHANDLE if needed.

jkotas commented on src/coreclr/src/interop/inc/interoplib.h:
```diff
@@ -0,0 +1,108 @@
+namespace InteropLib
+{
+    using OBJECTHANDLE = void*;
+    using OBJECTREF_PROTECTED = void*;
```
This makes me worried. This seems to be getting into manually managed code that VM has a lot of infrastructure for to make sure that it is written correctly.

If the interop library is cleanly separated, it needs to be 100% preemptive mode. 

If it needs to be manually managed code where something runs cooperative and something preemptive, etc., it needs to have the VM-style contract to capture and validate it.

AaronRobinsonMSFT replied:
I agree. I want to chat in person about the issue and hear your thoughts on what options exist. Avoiding this concept in the lib is possible, but does come with a cost. At the moment I was mitigating the risk by putting proper contracts on the call back into the runtime if this type is actually used.

jkotas replied:
Having contracts on the edges helps a bit, but it makes the code hard to review. It is the solution we have used in .NET Framework 1.0. We went through a lot of effort for .NET Framework 2.0 to put the contracts everywhere to make the code auditable.

AaronRobinsonMSFT replied:
> It is the solution we have used in .NET Framework 1.0. We went through a lot of effort for .NET Framework 2.0 to put the contracts everywhere to make the code auditable.

History wins again. I will take another shot at this.

AaronRobinsonMSFT replied:
@jkotas Thought on creating a function that the interoplib can call (i.e. assert) if the current thread is Cooperative? This would allow us to assert a contract in the interoplib. There are options, but keeping how to ensure a wrapper is active in the interoplib is my desire.

jkotas replied:
I do not think it is useful to create new style for writing manually managed code. I think that all manually managed code (ie what is running in cooperative mode) should be written in the same style.

AaronRobinsonMSFT replied:
Thanks for pointing this out @jkotas. Taking the time to rethink it was a good idea.

AaronRobinsonMSFT committed:
Fix Release builds

AaronRobinsonMSFT committed:
Apply review feedback about active wrapper API.

jkotas commented on src/coreclr/src/interop/comwrappers.h:
```diff
@@ -0,0 +1,264 @@
+#include <interoplib.h>
+#include "referencetrackertypes.h"
+
+enum class CreateComInterfaceFlagsEx : LONG
```
```suggestion
enum class CreateComInterfaceFlagsEx : int32_t
```

AaronRobinsonMSFT committed:
Update src/coreclr/src/interop/comwrappers.h

Co-Authored-By: Jan Kotas <jkotas@microsoft.com>

AaronRobinsonMSFT committed:
Bug in ignore cache scenario - found during local GCStress run.

AaronRobinsonMSFT committed:
Finish off minimal testing scenarios.

AaronRobinsonMSFT committed:
Fix build issues on macOS using clang.

AaronRobinsonMSFT committed:
Update Windows' build in response to macOS build fixes.

AaronRobinsonMSFT committed:
Release build fixes

Some CRST functions don't exist in Release builds.
Change usage of COMPlusThrow() to COMPlusThrowHR().

AaronRobinsonMSFT committed:
Remove ATL dependency from Mock Reference Tracker runtime library.

AaronRobinsonMSFT committed:
Always include ComHelper.h in COM tests.

AaronRobinsonMSFT committed:
Add Release() impl on test ComSmartPtr<T>.

AaronRobinsonMSFT committed:
Create a "support" and "nosupport" ComWrappers implementation.

AaronRobinsonMSFT committed:
Properly add a platform not supported implementation for ComWrappers.

AaronRobinsonMSFT committed:
Really remove the ComWrappers.NoSupport.cs file.

AaronRobinsonMSFT committed:
Forgot GetIUnknownImpl() API in the not supported version of ComWrappers.

AaronRobinsonMSFT committed:
Remove InteropLib shutdown logic.

jkotas commented on src/coreclr/src/vm/interoplibinterface.cpp:
```diff
@@ -0,0 +1,1249 @@
+        ~ExtObjCxtCache() = default;
+
+    public:
+        bool IsLockHeld()
```
Would be better to ifdef our the whole function be ifdefed for `_DEBUG` so that it does not get used by accident in release?

AaronRobinsonMSFT replied:
That is a fair argument. I didn't realize until recently that Crst only exposed the check in `DEBUG` mode. WIll move to enclose the function itself.

jkotas commented on src/coreclr/src/vm/interoplibinterface.cpp:
```diff
@@ -0,0 +1,1249 @@
+            // Shrink cache and clear unused handles.
+            refCache->ShrinkDependentHandles();
+        }
+        EX_CATCH
```
Throwing and catching exceptions when the runtime is suspect. I expect that you would see a assorted collection of assert/crashes if the exception was actually thrown/caught here.

AaronRobinsonMSFT replied:
Agreed. I originally didn't have the try/catch. I followed the existing RCW Walking algorithm and it had this try/catch. On top of that the path here indicates a `NOTHROW` environment so I was trying to make it similar. Technically the only throwing part of this code path is `GetAppDomain()->GetRCWRefCache()` since the `GetRCWRefCache()` lazily allocates an instance and can OOM.

I could simply just wrap that call and return if the OOM occurs similar to below:

```C++
if (nCondemnedGeneration >= 2)
{
    RCWRefCache* refCache;
    EX_TRY
    {
        refCache = GetAppDomain()->GetRCWRefCache();
    }
    EX_CATCH
    {
    }
    EX_END_CATCH(RethrowCorruptingExceptions)
    if (refCache == NULL)
        return;
    ExtObjCxtCache* cache = ExtObjCxtCache::GetInstanceNoThrow();

    ...
}
```

Alternatively, I could also handle the OOM entirely in the `GetRCWRefCache()` call and return `NULL`.

jkotas replied:
Why do we need to allocate RCW cache during the GC? If there is no RCW cache, there is nothing to do - no need to allocate one.

AaronRobinsonMSFT replied:
Not the cache, the ref cache is the issue. The ref cache is the mechanism that associates them - see https://github.com/dotnet/runtime/blob/master/src/coreclr/src/vm/rcwrefcache.cpp.

AaronRobinsonMSFT replied:
Just as an FYI, the plan is to update the `RCWRefCache` implementation when we remove the built-in WinRT support. I am simply reusing it here to avoid yet more code.

AaronRobinsonMSFT replied:
Actually how about I create the `RCWRefCache` at the same time I create the RCW Cache itself. This will allow me to remove the try/catch entirely.

jkotas replied:
Yes, that would be nice.

AaronRobinsonMSFT committed:
Acquire RefCache at the same time as creating ObjectCache. This avoids
  the need to handle a possible OOM at an inconvenient time.
Only define IsLockHeld() function in debug build.
Make usage of NULL consistent in interface cpp file.

AaronRobinsonMSFT committed:
Add StressLogs to InteropLibInterface

AaronRobinsonMSFT committed:
Cast OBJECTREFs so they work in StressLog macros.

AaronRobinsonMSFT committed:
Add a ComWrappers feature flag.

AaronRobinsonMSFT committed:
Remove duplicate entry.

AaronRobinsonMSFT commented:
All green - w00t! @jkotas  and @jkoritzinsky Any more feedback or thoughts? Still waiting for API review, but this is a large PR and will take some time. Thanks.

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.cs:
```diff
@@ -0,0 +1,224 @@
+        None = 0,
+
+        /// <summary>
+        /// The caller will provide an IUnknown Vtable.
```
Maybe describe the situation that this is mean to be used for? When does one want to use this flag?

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.cs:
```diff
@@ -0,0 +1,224 @@
+        /// <summary>
+        /// Ignore the internal cache when creating an object.
+        /// </summary>
+        IgnoreCache = 2,
```
Is this equivalent of `Marshal.GetUniqueObjectForIUnknown`? If yes, we should use "Unique" in this flags name - so that we use one name for the same concept.

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.cs:
```diff
@@ -0,0 +1,224 @@
+        /// This function can only be called a single time. Subsequent calls to this function will result
+        /// in a <see cref="System.InvalidOperationException"/> being thrown.
+        /// </remarks>
+        public void RegisterForReferenceTrackerHost()
```
I am wondering whether it would make sense to call this `RegisterAsGlobalInstance` - that is easier to understand name, and extend its meaning beyond reference tracking: The global instance can be also used to light-up the existing `Marshal` APIs.

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.cs:
```diff
@@ -0,0 +1,224 @@
+        /// <remarks>
+        /// The <paramref name="agileObjectRef"/> is an <see href="https://docs.microsoft.com/windows/win32/api/objidl/nn-objidl-iagilereference">IAgileReference</see> instance. This type should be used to ensure the associated external object, if not known to be free-threaded, is released from the correct COM apartment.
+        /// </remarks>
+        protected abstract object CreateObject(IntPtr externalComObject, IntPtr agileObjectRef, CreateObjectFlags flags);
```
Would it make sense to return null when this instance is not able to create the object? It would enable chaining of different ComWrappers together - if we ever need it.

jkotas replied:
(In either case, it would be useful to document the behavior when the object cannot be created.)

AaronRobinsonMSFT replied:
>It would enable chaining of different ComWrappers together - if we ever need it.

Not sure that scenario could be solved by returning `null`, but I can definitely document the current behavior.

jkotas replied:
> Not sure that scenario could be solved by returning null,

I though that it could be used to solve lighting up of the existing lighting-up of the existing `Marshal.GetObjectForIUnknown` and other APIs via this ComHelpers.

Basically the first thing that `Marshal.GetObjectForIUnknown` would do would be: if (global instance registered) { try to get the object via the global instance}, and then fallback to the existing built-in behavior.

jkotas replied:
(feel free to push back on this if you do not think that it is a good idea)

jkotas replied:
And do the same thing for `Marshal.GetIUnknownForObject` - by allowing ComputeVtable to return null.

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.cs:
```diff
@@ -0,0 +1,224 @@
+        {
+            if (instance == null)
+                throw new ArgumentNullException(nameof(instance));
+
```
Do we want to have any argument validation of the flags here? E.g. if `TrackerSupport` flag is specified, this instance have to be registered as global?

jkoritzinsky replied:
We need to allow specifying `TrackerSupport` even if it's not registered as global. That's the only way to efficiently do WinRT object activation.

More specifically, in the RCW case we need to be able to say "associate this pointer with this object" for a newly constructed object in that object's constructor. The way to do that in the ComWrappers API is to create a new ComWrappers-derived class and instance that just returns the given object and pass it the pointer.

jkotas replied:
I was under impression that the typical WinRT app is going to have just one ComWrapper instance, but I see that it is not the case.

Do you believe that you are doing for the RCW case is right, or would it make sense to have a more direct API for it?

jkoritzinsky replied:
Based on some of your other review comments, Aaron and I have chatted about using the idea of a context parameter for RCW creation to support the object activation scenario. A more direct API would definitely be nice, but I don’t believe it’s absolutely necessary.

jkotas replied:
If we realize in future that we need a more direct API, what would it look like?

I would like to test whether it can be naturally if needed, or whether it would be unnatural addition.

jkoritzinsky replied:
One option would be:

```
+ public object RegisterObjectForInterface(object obj, IntPtr externalComObject);
```

The API would add the mapping to the cache if it doesn't exist and then return the RCW for the COM object. In the activation scenario this would be the same object, but in the case of a race it returns the first one registered.

There's one example in CsWinRT where there can be a race: delegate marshalling. To avoid doing unneeded reflection (and because we can't determine the delegate's type without being told in advance), the delegate marshaling creates an RCW and then registers it instead of creating it within a ComWrappers instance. In the case of a race, we want to return only the registered RCW to the user.

jkotas replied:
I like this API. Should we just add it now? Should it take flags like the Get api?

jkoritzinsky replied:
I think we can add it now. And it should probably take flags so we can say "this needs tracker support".

jkoritzinsky replied:
I've added flags to my proposed addition above.

AaronRobinsonMSFT replied:
Can someone restate the need for this new API? It is basically the same as `GetOrCreateObjectForComInterface()` but always requires the creation of an `object`.

jkoritzinsky replied:
This API is for the scenario where you have an IntPtr and an object created outside of the ComWrappers and you want to tell the runtime “this object is the RCW for this pointer”. RuntimeClass activation and Delegate marshaling are the two scenarios where CSWinRT would use these. If it’s not provided, then CsWinRT would need to make additional subclasses of ComWrappers to implement it.

AaronRobinsonMSFT replied:
An alternative approach is to use the `ThreadStatic` approach mentioned below. The subclassing should happen anyways to optimize scenarios when look up isn't needed and the type is known. This API demonstrates an anti-pattern of always allocate and then possibly use. As opposed to the other API that defers allocation until required. The solution to the issue above is simply more code gen as opposed to creating an API that necessitates the anti-pattern of allocate first and then see if needed.

AaronRobinsonMSFT replied:
My perspective here is to avoid the creation of anti-pattern APIs if the solution can be accomplished through more code-gen. If code-gen can't accomplish the scenario then we should have the conversation.

AaronRobinsonMSFT replied:
After spending some more time with this I want to propose something that I think will address all concerns.

Instead of a new API that is yet another way to create an RCW, how about we change the `GetOrCreateObjectForComInterface()` to not take the `Type` instance as presently implemented but instead take an optional `object` which is the RCW. The semantics are the same, but if the `object` is provided the protected `CreateObject()` call will not be made since there is no need to create anything.

This keeps the semantics for creation clear - use `GetOrCreateObjectForComInterface()`, but allows for the annoying cases where the `object` was already created so there is no need to request creation.

Upon deeper reflection of this proposal my two biggest concerns are (1) How do consumers know to select this API versus the other? and (2) Making the API larger than needed. I get the problem and acknowledge it is real, I am simply not convinced this specific case warrants an entirely new function.

jkotas commented on src/coreclr/src/dlls/mscoree/coreclr/CMakeLists.txt:
```diff
@@ -105,6 +105,8 @@ set(CORECLR_LIBRARIES
     ildbsymlib
     utilcode
     libraries-native
+    v3binder
```
Incorrectly resolved merge conflict? v3binder is listed above.

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.cs:
```diff
@@ -0,0 +1,224 @@
+            => (comWrappersImpl ?? s_globalInstance!).ReleaseObjects(objects);
+
+        /// <summary>
+        /// Register this class's implementation to be used when a Reference Tracker Host instance is requested from another runtime.
```
"runtime" is very overloaded term. I would avoid mentioning it in the description. 

Maybe just say what the global instance can do that the local instances cannot:

- The wrappers on the global instance can be participate in object tracking via `CreateObjectFlags.TrackerObject/CreateComInterfaceFlags.TrackerSupport` flags

- (potentially if you do something about my other comment) The global instance is used to light-up existing Marshal APIs such as `Marshal.GetObjectForIUnknown`.

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.cs:
```diff
@@ -0,0 +1,224 @@
+        /// <remarks>
+        /// The <paramref name="agileObjectRef"/> is an <see href="https://docs.microsoft.com/windows/win32/api/objidl/nn-objidl-iagilereference">IAgileReference</see> instance. This type should be used to ensure the associated external object, if not known to be free-threaded, is released from the correct COM apartment.
+        /// </remarks>
+        protected abstract object CreateObject(IntPtr externalComObject, IntPtr agileObjectRef, CreateObjectFlags flags);
```
Should this take a desired `System.Type` of the wrapper as optional argument?

jkotas replied:
- To save lookup time when the wrapper type is known
- (potentially) To disambiguate the cases where there are multiple candidates

AaronRobinsonMSFT replied:
This suggestion aligns with something I had been considering with passing a "context". What are your thoughts about taking an optional `object` and the `ComWrappers` implementation could interpret that in an implementation dependent way.

AaronRobinsonMSFT replied:
This suggestion aligns with something I had been considering with passing a "context". What are your thoughts about taking an optional object and the ComWrappers implementation could interpret that in an implementation dependent way.

jkotas replied:
I got the idea for the `System.Type` argument by looking at existing System.Runtime.InteropServices.Marshal APIs, and validating whether a similar expressiveness can be achieved through this.

System.Object - I am not sure. If people need to smuggle extra arbitrary data, it can be done via ThreadStatic.

jkotas replied:
Or is the scenario where the wrapper type is known covered by the RegisterObjectForInterface discussed in the other comment?

AaronRobinsonMSFT replied:
I am not convinced of the need/use for the `RegisterObjectForInterface()`. I don't believe it has a use since it basically requires creation of an object that may be useless. In my opinion supplying the `Type` in this API makes much more sense.

AaronRobinsonMSFT replied:
@jkotas What is the typical response for default parameters in new APIs? I would like to avoid users from having to provide `null`, but realize default parameters aren't always desirable. Also, see [here](https://github.com/dotnet/runtime/pull/32091#discussion_r382956773) for changing the `Type?` to `object?`.

jkotas replied:
https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/member-overloading says "DO use member overloading rather than defining members with default arguments". This is guideline from 2008. I believe that the current API design trend is more amicable to use of default arguments than this. I do not think there is a simple rule.

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.cs:
```diff
@@ -0,0 +1,224 @@
+            public Guid IID;
+
+            /// <summary>
+            /// Must be pinned memory that is owned by the implementer of <see cref="ComputeVtables(object, CreateComInterfaceFlags, out int)"/>.
```
This should say the same thing as https://github.com/dotnet/runtime/pull/32091/files#diff-0ad06c3fa56ab2b9ed0f7c33f97914a0R135 . Maybe just say here the lifetime must be the same as the lifetime of the ComInterfaceEntry?

AaronRobinsonMSFT committed:
Review feedback.

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.cs:
```diff
@@ -0,0 +1,235 @@
+        /// <remarks>
+        /// The <paramref name="agileObjectRef"/> is an <see href="https://docs.microsoft.com/windows/win32/api/objidl/nn-objidl-iagilereference">IAgileReference</see> instance. This type should be used to ensure the associated external object, if not known to be free-threaded, is released from the correct COM apartment.
+        ///
+        /// If the object cannot be created and <code>null</code> is returned, the call to <see cref="ComWrappers.GetOrCreateObjectForComInstance(IntPtr, CreateObjectFlags)"/> will throw a <see cref="System.ArgumentNullException"/>.
```
If we allow `null` to be returned, the return type signature should be `object?`

AaronRobinsonMSFT committed:
Add Type? to the GetOrCreateObjectForComInstance() API call.

AaronRobinsonMSFT committed:
Revert "Add Type? to the GetOrCreateObjectForComInstance() API call."

This reverts commit d0f6a5b04adbaa8e53445e4474780a9182ff8bde.

AaronRobinsonMSFT committed:
Update GetOrCreateObjectForComInstance() to accept 'object?' as the
 defined wrapper. This will indicate to skip calling CallObject().

AaronRobinsonMSFT committed:
Validate failure for reuse of wrapper for external object.

AaronRobinsonMSFT committed:
Fix bug in failure case of attempting to reuse a wrapper.

AaronRobinsonMSFT committed:
Remove IAgileReference usage.

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.cs:
```diff
@@ -0,0 +1,242 @@
+        ///
+        /// If the <paramref name="wrapper"/> instance already has an associated external object a <see cref="System.NotSupportedException"/> will be thrown.
+        /// </remarks>
+        public object GetOrCreateObjectForComInstance(IntPtr externalComObject, CreateObjectFlags flags, object? wrapper = null)
```
As we have discussed offline, it is unusual to have all possible Add and Create variants bundled into a single API. E.g. compare it with design of methods on ConcurrentDictionary.

This is a question for API review to answer.

jkotas commented:
> Marshal.GetIUnknownForObject() - Calling this on a wrapper object created by this new API will passively create an old style CCW. Deferring any changes until customer reported.

This should be easily doable now that we have made `RegisterAsGlobalInstance`. Now sure whether it still make sense to defer or whether it would make more sense to do it now.

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.cs:
```diff
@@ -0,0 +1,242 @@
+        /// </remarks>
+        protected virtual void ReleaseObjects(IEnumerable objects)
+        {
+            throw new NotImplementedException();
```
Would it make more sense for the default impl to be no-op?

AaronRobinsonMSFT replied:
I thought about that. My feeling was a request for releasing objects isn't something that should be ignorable. If there is nothing to do be done then implementers can state that explicitly. I want implementers to be put into a position to have to make decisions.

jkotas replied:
The typical API design preference is to avoid forcing decision like this by providing a reasonable default implementation.

The class may not even be abstract. We can make the two abstract methods to have a reasonable default implementations. 

Something to ask during API review...

jkotas reviewed (approved):
It would be nice if Jeremy can review this as well.

AaronRobinsonMSFT committed:
Apply API review comments.

AaronRobinsonMSFT committed:
Update test to align with API changes.

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.cs:
```diff
@@ -0,0 +1,249 @@
+            return GetOrCreateObjectForComInstanceEx(externalComObject, flags, wrapper);
+        }
+
+        private object GetOrCreateObjectForComInstanceEx(IntPtr externalComObject, CreateObjectFlags flags, object? wrapper)
```
Nit: This should be called `GetOrCreateObjectForComInstance` or `GetOrCreateObjectForComInstanceInternal`.

AaronRobinsonMSFT replied:
@jkotas Is the `Ex` suffix no good?

jkotas replied:
It feels very Windowsy and it does not follow the usual managed naming conventions. I do not think we have any method with `Ex` suffix in CoreLib, except for the Windows API interop methods.

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.cs:
```diff
@@ -0,0 +1,249 @@
+            if (externalComObject == IntPtr.Zero)
+                throw new ArgumentNullException(nameof(externalComObject));
+
+            ComWrappers impl = this;
```
I expect that this will be pretty hot method and that we are going to make two leaner dedicated paths for this once we put it under profiler...

jkotas replied:
Related: It would be interesting to see how fast is this vs. regular built-in COM interop.

AaronRobinsonMSFT replied:
What is the "leaner" path? There is a single if check.

jkotas replied:
You are passing the extra argument around, etc. It adds overhead. I am just speculating about the perf problems. Actually measuring the perf and comparing it with the build-in would be the best. I hope that this can match it.

AaronRobinsonMSFT replied:
Okay. That will be done during the feedback phase with CsWinRT.

AaronRobinsonMSFT replied:
I did run this under a profiler though and this was never even close to being hot. The mere act of allocating was more expensive.

jkotas replied:
I expect that the common path through this is going to be lookup of existing object (ie no allocation).

> the feedback phase with CsWinRT.

You do not need to wait for that. You can just make a few microbenchmarks to compare built-in COM interop with this.

AaronRobinsonMSFT replied:
>You do not need to wait for that. You can just make a few microbenchmarks to compare built-in COM interop with this.

Since there is close to no overlap between the two activation/creation scenarios I am going to assume the microbenchmarks in question would relate to re-querying the system for the associated RCW/CCW?. The equivalent to these in the built-in system would be [`Marshal.GetIUnknownForObject()`](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.marshal.getiunknownforobject) and [`Marshal.GetObjectForIUnknown()`](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.marshal.getobjectforiunknown) - are those the APIs you were thinking about for comparison benchmarking?

jkotas replied:
Yep

AaronRobinsonMSFT replied:
@jkotas Here are the basic results from the two different APIs. I ran this in the opposite order with the same results. The `COMPlus_TieredCompilation` option had no noticeable impact.

Iteration count: 1000000

```
GetOrCreateComInterfaceForObject: - 35.8286 ns/iter
GetIUnknownForObject: - 54.6582 ns/iter
GetOrCreateComInterfaceForObject: - 37.142599999999995 ns/iter
GetIUnknownForObject: - 55.4512 ns/iter
```

```
GetOrCreateObjectForComInstance: - 64.417 ns/iter
GetObjectForIUnknown: - 86.2446 ns/iter
GetOrCreateObjectForComInstance: - 67.9225 ns/iter
GetObjectForIUnknown: - 85.1364 ns/iter
```

jkotas replied:
The numbers for `GetOrCreateComInterfaceForObject` look suspiciously good. What is the lifetime contract for `GetOrCreateComInterfaceForObject` ? Does it return AddRefed IUnknown like GetIUnknownForObject?

AaronRobinsonMSFT replied:
> The numbers for GetOrCreateComInterfaceForObject look suspiciously good

Yes they do. I am looking at this in a profiler and attempting to ascertain why.

>Does it return AddRefed IUnknown like GetIUnknownForObject?

Yes.

AaronRobinsonMSFT replied:
From the looks of the trace the issue here is the scenario I chose hit a degenerate case where the system must do a lot more work to query the associated IClassX tables and an aggregation check. This results in at least two more AddRef()/Release() cycles. Factoring that out, the results become almost identical. The new approach is still slightly faster, but that seems to be because the new code to get at the `IUnknown` is just simpler/more direct.

jkotas replied:
Ok, nice!

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.cs:
```diff
@@ -193,17 +178,39 @@ public object GetOrCreateObjectForComInstance(IntPtr externalComObject, CreateOb
+        /// If the <paramref name="wrapper"/> instance already has an associated external object a <see cref="System.NotSupportedException"/> will be thrown.
         /// </remarks>
-        protected virtual void ReleaseObjects(IEnumerable objects)
+        public object GetOrRegisterObjectForComInstance(IntPtr externalComObject, CreateObjectFlags flags, object wrapper)
```
Should this check that the wrapper is non-null?

AaronRobinsonMSFT replied:
Boo. Forgot that the enforcement is compiler not functional. Yes it should.

AaronRobinsonMSFT committed:
Merge conflict.

AaronRobinsonMSFT committed:
Review feedback.

jkoritzinsky commented:
It looks like you need to add stub/throwing definitions of the new API surface in Mono's System.Private.CoreLib.

AaronRobinsonMSFT committed:
Add stub for AllocateTypeAssociatedMemory() in Mono build.

AaronRobinsonMSFT committed:
Style and name.

jkotas commented on src/coreclr/tests/src/baseservices/compilerservices/RuntimeHelpers/AllocateTypeAssociatedMemory.cs:
```diff
@@ -0,0 +1,43 @@
+
+class AllocateTypeAssociatedMemoryTest
+{
+    private static void ValidateInvalidArguments()
```
This is a regular API test. It would be better for it to live under src\libraries

AaronRobinsonMSFT replied:
Do you have a suggestion on where to put this? I don't know what project to add this to and honestly don't really want to figure out how to add something new to the libraries partition.

AaronRobinsonMSFT replied:
Sorry nevermind. I found it.

jkotas replied:
https://github.com/dotnet/runtime/blob/master/src/libraries/System.Runtime/tests/System/Runtime/CompilerServices/RuntimeHelpersTests.cs

AaronRobinsonMSFT committed:
Move RuntimeHelpers.AllocateTypeAssociatedMemory tests to src/libraries.

AaronRobinsonMSFT committed:
Missed static

AaronRobinsonMSFT committed:
Skip tests on Mono.

AaronRobinsonMSFT committed:
Unit test fix for RuntimeHelpers.AllocateTypeAssociatedMemory().

AaronRobinsonMSFT committed:
Bad merge on rebase.

AaronRobinsonMSFT committed:
Minor rewording for the interoplib interface.

jkotas commented on src/coreclr/src/interop/inc/interoplibimports.h:
```diff
@@ -0,0 +1,78 @@
+    enum class GcRequest
+    {
+        Default,
+        Extensive // This is an expensive GC request, akin to a Gen2/"stop the world" GC.
```
"extensive GC" is not a term I have ever heard in connection with GC.

This does full blocking GC. Should this be called `FullBlocking`?

AaronRobinsonMSFT replied:
Yep. I was struggling with a name. I can do `FullBlocking`. Will see if this is all green first and then change it.

AaronRobinsonMSFT committed:
Feedback on naming.
